projects/project-2/readings/Project 2 - SPARQL library

```diff
@@ -0,0 +1,297 @@
+Constraint Description: Identifies equivalent properties with different range specifications.
+Severity: Error
+
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+SELECT ?property1 ?property2 ?range1 ?range2
+WHERE {
+  ?property1 a owl:ObjectProperty .
+  ?property2 a owl:ObjectProperty .
+  ?property1 owl:equivalentProperty ?property2 .
+  ?property1 rdfs:range ?range1 .
+  ?property2 rdfs:range ?range2 .
+  FILTER(?range1 != ?range2)
+}
+
+The goal of this query is to identify pairs of object properties that are defined as equivalent but have different specified ranges.​ 
+This discrepancy can indicate potential issues in the ontology's design, as equivalent properties should maintain consistent range definitions to ensure accurate data interpretation.
+
+8. Analyzing Class Hierarchies for Inconsistent Type Definitions
+
+Title: Inconsistent Class Type Definition
+Constraint Description: Detects classes assigned multiple, different types.
+Severity: Error
+
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+SELECT ?class ?type1 ?type2
+WHERE {
+  ?class a ?type1 .
+  ?class a ?type2 .
+  FILTER(?type1 != ?type2)
+  FILTER(?type1 != rdfs:Class && ?type2 != rdfs:Class)
+}
+
+The primary objective of this query is to identify classes assigned multiple types in an ontology that differ from each other.​ 
+Inconsistent type definitions can create ambiguity regarding the intended semantics of the classes and lead to erroneous interpretations in data retrieval and reasoning.
+
+9. Detecting Potentially Misplaced Inverse Object Properties (Level 5 - 5 points)
+
+Title: Unused Inverse Property Detection
+Constraint Description: Identifies inverse object properties without corresponding instance relationships.
+Severity: Warning
+
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+SELECT ?property1 ?property2
+WHERE {
+  ?property1 a owl:ObjectProperty .
+  ?property2 a owl:ObjectProperty .
+  ?property1 owl:inverseOf ?property2 .
+  FILTER NOT EXISTS {
+    ?instance1 ?property1 ?instance2 .
+    ?instance2 ?property2 ?instance1 .
+  }
+}
+
+This query object properties defined as inverses within the ontology without corresponding instances illustrating those relationships. 
+This lack of practical implementation can signify potential design flaws within the ontology framework.
+
+10. Identifying Classes with No Properties (Level 4 - 10 points)
+
+Title: Empty Class Property Detection
+Constraint Description: Detects classes that have no properties defined.
+Severity: Warning
+
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+SELECT ?class
+WHERE {
+  ?class a rdfs:Class .
+  FILTER NOT EXISTS {
+    ?class ?property ?value .
+    FILTER(?property != rdf:type)
+  }
+}
+
```

Comment:
I think this is nice, but can be improved by including other types of class definitions like owl:Class. One way to do this is to expand the ?class a rdfs:Class part to include owl:Class or other relevant definitions. Hope this helps!

Reply:
Thank you, I'll try!